lib/ignore/ignore_test.go

```diff
@@ -1098,3 +1098,37 @@ func TestPartialIncludeLine(t *testing.T) {
 		}
 	}
 }
+
+func TestSkipIgnoredDirs(t *testing.T) {
+	tcs := []struct {
+		pattern  string
+		expected bool
+	}{
+		{`!/test`, true},
+		{`!/t[eih]t`, true},
+		{`!/t*t`, true},
+		{`!/parent/test`, true},
+		{`!/parent/t[eih]t`, true},
+		{`!/parent/t*t`, true},
+		{`!/pa\[esd\]rent/t*t`, true},
+		{`!/pa*nt/test`, false},
+		{`!/pa[sdf]nt/t[eih]t`, false},
+		{`!/lowest/pa[sdf]nt/test`, false},
+		{`!/lo*st/parent/test`, false},
+		{`/pa*nt/test`, true},
```

Comment:
I am not sure why some cases have the outcome that they have.

I'd also like to see some tests that cover multiple patterns proving the scans don't dvelve down into ignored dirs.

Reply:
After digging into the code I think I now understand why it is the way it is, but I'd still like a wider test, for example `!/**.mp3`, last component would not be checked, yet double star allows you to punch through all levels.

Reply:
Thanks so much. I missed both that case and another complication (I knew there were nasty pitfalls here...):  
When a user adds the pattern `!/foo*` we internally add `!/foo*` and `!/foo*/**`. Under my current rules this results in not skipping ignored dirs, which is unnecessary. However if a user enters `!/foo*/` or `!/foo*/**`, ignored dirs should indeed not be skipped. However the two cases are indistinguishable when looking at a single `Pattern`.

So either this gets more complex. Or we just optimize for rooted paths without any wildcards anywhere.

Reply:
I found a way to do it with little complexity. Maybe a bit hacky, but what isn't when it comes to ignores. I also added more tests, include one in lib/scanner to check that we really don't delve.